specs/HubProtocol.md

```diff
@@ -0,0 +1,225 @@
+
+On the Caller side, the user code which performs the invocation indicates how it would like to receive the results and it is up the Caller's Binder to determine how to handle the result. If the Caller expects only a single result, but multiple results are returned, the Caller's Binder should yield an error indicating that multiple results were returned. However, if a Caller expects multiple results, but only a single result is returned, the Caller's Binder should yield that single result and indicate there are no further results.
+
+## JSON Encoding
+
+In the JSON Encoding of the SignalR Protocol, each Message is represented as a single JSON object, which should be the only content of the underlying message from the Transport.
+
+### Invocation Message Encoding
+
+An `Invocation` message is a JSON object with the following properties:
+
+* `type` - A `Number` with the literal value 1, indicating that this message is an Invocation.
+* `invocationId` - A `String` encoding the `Invocation ID` for a message.
+* `target` - A `String` encoding the `Target` name, as expected by the Callee's Binder
+* `arguments` - An `Array` containing arguments to apply to the method referred to in Target. This is a sequence of JSON `Token`s, encoded as indicated below in the "JSON Payload Encoding" section
+
+Example:
+
+```json
+{
+    "type": 1,
+    "invocationId": 123,
+    "target": "Send",
+    "arguments": [
+        42,
+        "Test Message"
+    ]
+}
+```
+
+### Result Message Encoding
+
+A `Result` message is a JSON object with the following properties:
+
+* `type` - A `Number` with the literal value 2, indicating that this message is a Result.
+* `invocationId` - A `String` encoding the `Invocation ID` for a message.
+* `error` - A `String` encoding the error message. The presence of this field indicates that an error should be yielded, **even if the payload field is also present**
+* `payload` - A `Token` encoding the result value (see "JSON Payload Encoding" for details). If error is present, this field is **completely ignored**.
+
+Example - A `Result` with a payload.
+
+```json
+{
+    "type": 2,
+    "invocationId": 123,
+    "payload": 42
+}
+```
+
+Example - A `Result` with an error.
+
+```json
+{
+    "type": 2,
+    "invocationId": 123,
+    "error": "An error occurred!"
+}
+```
+
+Example - Even though there is a `payload`, the presence of `error` means it is ignored and the error is yielded. This should never be transmitted by a conforming endpoint, but endpoints should be prepared to receive messages of this format, to be robust.
+
+```json
+{
+    "type": 2,
+    "invocationId": 123,
+    "error": "An error occurred!",
+    "payload": 42
+}
+```
+
+### Completion Message Encoding
+
+A `Completion` message is a JSON object with the following properties
```

Comment:
I think having just one message for single results will be less error prone and prevent from cases where result was received but the callback never finished because the Completion message never arrived. To be clear we would still have both Result and Completion messages but Result messages will not be used for singleton results.